### ISSUE TITLE
twisted.names shouldn't stop listening when crappy data is sent

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11193]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#1708|
|Type|enhancement|
|Created|2006-05-08 14:38:40Z|

It should catch the exception and log it.

```
2006/04/30 19:10 CDT [twisted.names.dns.DNSDatagramProtocol (UDP)] Traceback (most recent call last):
          File "/home/t-names/Twisted/twisted/python/log.py", line 53, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/home/t-names/Twisted/twisted/python/log.py", line 38, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/home/t-names/Twisted/twisted/python/context.py", line 59, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/home/t-names/Twisted/twisted/python/context.py", line 37, in callWithContext
            return func(*args,**kw)
        --- <exception caught here> ---
          File "/home/t-names/Twisted/twisted/internet/selectreactor.py", line 139, in _doReadOrWrite
            why = getattr(selectable, method)()
          File "/home/t-names/Twisted/twisted/internet/udp.py", line 126, in doRead
            self.protocol.datagramReceived(data, addr)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 1088, in datagramReceived
            m.fromStr(data)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 1047, in fromStr
            self.decode(strio)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 992, in decode
            header = readPrecisely(strio, self.headerSize)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 153, in readPrecisely
            raise EOFError
        exceptions.EOFError:
 
2006/04/30 19:10 CDT [twisted.names.dns.DNSDatagramProtocol (UDP)] (Port 53 Closed)
2006/04/30 19:10 CDT [twisted.names.dns.DNSDatagramProtocol (UDP)] Stopping protocol <twisted.names.dns.DNSDatagramProtocol instance at 0xb798580c>
```

<details><summary>Searchable metadata</summary>

```
trac-id__1708 1708
type__enhancement enhancement
reporter__radix radix
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1147099120000000 1147099120000000
changetime__1149638177000000 1149638177000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

Is this fixed now, or do you want separate t.names fix too?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

It should probably be fixed in names too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

Fixed in [branches/names-eof-1708](branches/names-eof-1708) - please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jerub|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @exarkun|
|-|-|

Tests pass, the change is a good change to make. I'm a little about there being two error paths, and only one being tested. Please add tests for 'Message.fromStr', testing that object in isolation. Verify that it decodes a DNS packet, and verify that it throws an EOFError error on truncated packets. Knowing what other exceptions can be thrown (instead of using a bare 'except:') would be nice, but not entirely necessery.

The API for Message() looks fugly here. This pattern:
```
m = Message()
m.fromStr(mystr)
```
is better expressed by implementing 'fromStr' as a classmethod, making the api:
```
m = Message.fromStr(mystr)
```
The function would go from looking like:
```
    def fromStr(self, str):     
        strio = StringIO.StringIO(str)
        self.decode(strio)
```
to looking like:
```
    def fromStr(cls, str):     
        strio = StringIO.StringIO(str)
        self = cls()
        self.decode(strio)
        return self
    fromStr = classmethod(fromStr)
```
This is a reasoanbly large change. Only persue this if you think it's worth doing. Create a new ticket maybe?

pyflakes really hates dns.py. We use exec when importing in order to insert query types into the module namespace. This is yuck, but I don't consider it an issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Added a real unit test for Message.fromStr behavior.  Didn't add any test for the non-EOFError exception branch, since no code should actually trigger that (and afaik none does).

fromStr predates classmethod, which is why it isn't one.  Changing it to one would be incompatible, although I can imagine another kind of descriptor as which it would retain compatibility while also acting as a classmethod.  It's mostly internal though so I'm not too worried about it.

The exec's in dns.py should really go away, I guess.  I'll make a ticket for that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [17096]) Merge names-eof-1708

Author: exarkun
Reviewer: Jerub
Fixes [#11193](https://github.com/twisted/twisted/issues/11193)

This adds several simple tests for the case of receiving a truncated DNS
messages and fixes the code so that it does not raise an exception all the
way to the reactor for these cases.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
